app/matlab_reader.h

```diff
@@ -0,0 +1,166 @@
+/// \file
+/// Maintainer: Felice Serena
+///
+///
+
+#pragma once
+
+#include "generic/types.h"
+#include "generic/disparity_map.h"
+#include "generic/frame.h"
+#include <set>
+#include <map>
+#include <Eigen/Core>
+#include <boost/filesystem.hpp>
+
+namespace MouseTrack {
+
+namespace fs = boost::filesystem;
+
+
+/// The MatlabReader is a very specialized class, that reads preprocessed disparity maps and
+/// Parameter files as created by the `extract_camera_streams.m` script.
+/// The expected file format and folder structures are hard coded.
+/// Takes the path to a folder corresponding to a bag file (via constructor) and gives
+/// access to the data via its methods.
+///
+/// Convention for methods: always first stream, then frame number.
+class MatlabReader {
+public:
+    /// This is the constructor you should use, give it the root to the bag file directory
+    MatlabReader(fs::path root_directory);
+
+    /// true if constructor decided, the file system is formatted properly, false otherwise
+    bool valid() const;
+
+    /// Index of first existing stream
+    StreamNumber beginStream() const;
+
+    /// C++ convention: points to the first non existing stream at the end
+    StreamNumber endStream() const;
+
+    /// index of first existing frame
+    FrameNumber beginFrame() const;
+
+    /// Index of first non existing frame at the end
+    FrameNumber endFrame() const;
+
+    /// A list of all paths (files, directories, ...) that were skipped during the preflight
+    const std::vector<fs::path>& ignoredPaths() const;
+
+    /// A list of all channels that are ignored (files that match the regular expression
+    /// for a sequence of stream images, but has an unknown base name).
+    const std::vector<std::string>& ignoredChannels() const;
+
+    /// Fetch normalized disparity map for frame f and stream s
+    DisparityMap normalizedDisparityMap(StreamNumber s, FrameNumber f) const;
+
+    /// Fetch disparity map as stored in bag file for frame f and stream s
+    DisparityMap rawDisparityMap(StreamNumber s, FrameNumber f) const;
+
+    /// Fetch channel parameters (focallength, baseline, ccx, ccy)
+    Eigen::MatrixXd channelParameters(FrameNumber f) const;
+
+    /// Fetch left camera picture for frame f and stream s
+    Picture picture(StreamNumber s, FrameNumber f) const;
+
+    /// Fetch rotation corrections for cameras
+    std::vector<Eigen::Matrix4d> rotationCorrections() const;
+
+    /// Fetch the camera chain (8 rotation matrices that define how the cameras are positioned to each other)
+    std::vector<Eigen::Matrix4d> camchain() const;
+
+    /// Fetch all data belonging to frame f of and all streams
+    std::vector<Frame> frames(FrameNumber f) const;
```

Comment:
This return type is probably wrong.

- There are things depending on the frame and the stream.
- There are things depending on the frame index (equal for all frames with same index)
- There are things depending on the stream.
- There are things that depend on nothing. 

We do want to denormalize a bit. We probably need a construct like a "FrameStack" or "FrameWindow" which holds parameters constant for a frameIndex and also a list of "Frame" with all parameters differing between frames.

We might need to define the terminology at some point.

Reply:
I think I should go to bed. If we assume a camchain to be defined according to the stream indexes, this actually should work, since each Frame holds one camchain matrix and one rotation corrections. I'll think about it tomorrow.

Reply:
I've made my mind up, to be consistent with other concepts like "Cluster" an object like "FrameWindow" makes sense. It also gives the flexibility to add values constant accross frames with the same index.